### ISSUE TITLE
Transaction Name is changing after I set it if called from another instrumented program

### ISSUE THREAD

**Issue by willie** [newrelic/go-agent#310]
## Description

I have 2 Go programs: Alice and Bob. Alice and Bob are instrumented with the New Relic agent on the same account.

Direct requests to Bob look like: `/header/L-999-70000.vbk`, but sometimes Alice calls Bob for the same request. I'm getting different transaction names logged if Bob is called directly vs. if Alice calls Bob instead.

My middleware calls `txn.SetName()` after the New Relic handler runs ServeHTTP and normally logs the handler pattern I registered in Chi. I'm using `newrelic.NewRoundTripper` to set up the default transport.

tl;dr Calls directly made to Bob show up in APM transactions as intended, but calls to Bob made from another New Relic app ignore my call to `txn.SetName()`

## Steps to Reproduce

I wrote a middleware for New Relic, registered after NR, but before my handler. It looks like this:

```
func NewRelic(app *newrelic.Application, options ...func(*newRelicConfig)) func(http.Handler) http.Handler {
	nr := &newRelicConfig{}
	for _, option := range options {
		option(nr)
	}

	return func(next http.Handler) http.Handler {
		_, handler := newrelic.WrapHandle(app, nr.pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if nr.pattern == "" {
				txn := newrelic.FromContext(r.Context())
				txn.SetName(r.Method + " " + r.URL.Path)
			}

			next.ServeHTTP(w, r)
		}))
		return handler
	}
}
```

And another handler called after that updates it with Chi information looks like this:

```
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			log.Println("hello")
			next.ServeHTTP(w, r)

			txn := newrelic.FromContext(r.Context())
			if txn == nil {
				return
			}

			ctx := chi.RouteContext(r.Context())
			if ctx == nil {
				return
			}

			// update the pattern
			pattern := ctx.RoutePattern()
			txn.SetName(r.Method + " " + pattern)
```

My `RoutePattern()` is `"/header/{vbk}"`

If I call Bob directly, New Relic will log a call to `/header/L-999-70000.vbk` as `/header/{vbk}` as an APM transaction.

If Alice calls Bob as part of handling another request, New Relic will log a call to `/header/L-999-70000.vbk` as `/header/L-999-70000.vbk` as an APM transaction.

Diagnostic logging shows that my handler is being called in the same way each time and that my `ctx.RoutePattern()` is returning the same information each time.

## Expected Behavior

I expect `txn.SetName()` to work the same for a direct call or one that has being made cross application. Direct calls are correctly grouped by pattern in APM, Cross application calls show up with the exact path in Bob, but have the correct pattern in Alice.

## Your Environment

Go 1.16.3 and go.mod has `github.com/newrelic/go-agent/v3 v3.11.0`
 
## Reproduction case

I can hack one together if this doesn't ring any bells.

**Comment by RichVanderwal** [comment]
Hi @willie , thanks for writing in! This doesn't ring any bells, so to help us figure this out for you, we have some questions to help us narrow this down:
* Do you have Distributed Tracing turned on? Are you using an older version cross-application tracing?
* When you view debug agent logs from Bob, is there a difference in processing between each type of request you describe? If you'd like to send us logs, you should open up an official support ticket by clicking the "Agents and installation" button on our [Support page](https://support.newrelic.com).

Our best guess right now is that either your txn.SetName() isn't taking effect in calls from Alice, or something else in the Go Agent is _**re**_-renaming the transaction after you're done naming it.

**Comment by willie** [comment]
I'm not sure the answer to either of those questions actually LOL, but I did go ahead and put together a reproducible test case.

https://github.com/willie/newrelic-transaction-name-bug

**Comment by willie** [comment]
<img width="328" alt="Screen Shot 2021-05-07 at 1 51 57 PM" src="https://user-images.githubusercontent.com/20665/117496012-09aa8200-af3c-11eb-93ad-bbeff6036b6d.png">

This is what it looks like for poor ol' bob.

**Comment by willie** [comment]
Were you able to reproduce with the test case I put together?

**Comment by RichVanderwal** [comment]
Hi @willie , we were able to reproduce your problem today. Thanks for that easy example, it really helped us troubleshoot this for you!

What you're seeing is an artifact of our old Cross-Application Tracing feature -- it "freezes" the transaction name earlier than normal so your two applications can be sure they're talking about the same transaction. This is happening before your `txn.SetName()` call, rendering it useless.

We have two alternatives for you:
1. Turn on [Distributed Tracing](https://docs.newrelic.com/docs/distributed-tracing/). This turns off the old Cross-Application Tracing feature, replacing it with a new system that doesn't do the same sort of transaction name freezing. We've tried it with your example and it all works as expected. Just add this to your configuration options when you're calling `newrelic.NewApplication()`:

```
		newrelic.ConfigDistributedTracerEnabled(true),
```
We recommend this approach! In fact, we're currently working on turning on Distributed Tracing by default for all of our language agents, leaving the old Cross-Application Tracing system behind.

2. Turn off Cross-Application Tracing. This will keep your applications from trading header information that lets New Relic correlate between your two applications. In the same `newrelic.NewApplication()` call:

```
		func(config *newrelic.Config) {
			config.CrossApplicationTracer.Enabled = false
		},
```

We hope this helps! Let us know if you have any comments or questions.
